code/code/misc/discord.cc

```diff
@@ -16,97 +19,137 @@ namespace po = boost::program_options;
+    "see discord.h")("achieve_threshold",
+    po::value<int>(&ACHIEVEMENT_THRESHOLD)->default_value(80), "see discord.h");
+
+  po::options_description config_options;
+  config_options.add(configOnly);
+
+  po::variables_map vm;
+  po::notify(vm);
+  std::ifstream ifs(configFile.c_str());
+
+  if (!ifs.is_open()) {
+    vlogf(LOG_MISC,
+      format("Discord webhooks: Failed to open config file '%s'") % configFile);
+    vlogf(LOG_MISC,
+      format("Discord webhooks: ensure config is located in: '%s'") %
+        std::filesystem::current_path());
+    return false;
+  }
+
+  po::store(parse_config_file(ifs, config_options), vm);
+  po::notify(vm);
+
+  return true;
 }
 
-
 // send a message to a discord webhook
 // we use the curl library for this and keep it simple
-bool Discord::sendMessageAsync(sstring channel, sstring msg)
-{
-
-    CURL* curl = curl_easy_init();
-    if (!curl) {
-        vlogf(LOG_MISC, "Discord webhooks: curl_easy_init() failed");
-        return false;
-    }
+bool Discord::sendMessageAsync(sstring channel, sstring msg) {
+  CURL* curl = curl_easy_init();
+  if (!curl) {
+    vlogf(LOG_MISC, "Discord webhooks: curl_easy_init() failed");
+    return false;
+  }
+
+  // sanitize and format the message
+  msg = format("{\"content\": \"%s\"}") % stripColorCodes(msg).escapeJson();
+
+  const char* webhookURL = channel.c_str();
+  const char* content = msg.c_str();
+
+  struct curl_slist* headers = NULL;
+  headers = curl_slist_append(headers, "Content-Type: application/json");
+
+  curl_easy_setopt(curl, CURLOPT_URL, webhookURL);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+  curl_easy_setopt(curl, CURLOPT_POSTFIELDS, content);
+
+  CURLcode res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
+    vlogf(LOG_MISC,
+      format("Discord webhooks: curl_easy_perform() failed: '%s'") %
+        curl_easy_strerror(res));
+  }
+
+  curl_slist_free_all(headers);
+  curl_easy_cleanup(curl);
+  return true;
+}
 
-    // sanitize and format the message
-    msg = format("{\"content\": \"%s\"}") % stripColorCodes(msg).escapeJson();
+void Discord::sendMessage(sstring channel, sstring msg) {
+  if (channel == "") {
+    // no channel configuration, so bail
+    return;
+  }
 
-    const char* webhookURL = channel.c_str();
-    const char* content = msg.c_str();
+  vlogf(LOG_MISC, format("Discord webhooks add to thread: '%s'") % msg);
 
-    struct curl_slist* headers = NULL;
-    headers = curl_slist_append(headers, "Content-Type: application/json");
+  std::lock_guard<std::mutex> lock(queue_mutex);
+  message_queue.push({channel, msg});
+}
 
-    curl_easy_setopt(curl, CURLOPT_URL, webhookURL);
-    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
-    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, content);
+void Discord::messenger() {
+  while (true) {
+    std::lock_guard<std::mutex> lock(queue_mutex);
+    if (message_queue.empty() && !stop_thread) {
```

Comment:
Busy loop, consuming 100% of available CPU time. Should use some sort of an event system instead.

Reply:
Perhaps you could `wait` for an event. I haven't done it in C++, so I don't know the very best way of doing this, but https://en.cppreference.com/w/cpp/atomic/atomic_wait looks promising.

Reply:
Ended up using condition_variable for this, which seems to work nicely.